src/H5Dmpio.c

```diff
@@ -4084,20 +4277,86 @@ H5D__mpio_collective_filtered_chunk_read(H5D_filtered_collective_io_info_t *chun
     (void)mpi_rank;
 #endif
 
-    if (chunk_list->num_chunk_infos) {
-        /* Retrieve filter settings from API context */
-        if (H5CX_get_err_detect(&err_detect) < 0)
-            HGOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "can't get error detection info");
-        if (H5CX_get_filter_cb(&filter_cb) < 0)
-            HGOTO_ERROR(H5E_DATASET, H5E_CANTGET, FAIL, "can't get I/O filter callback function");
+    /* Cache some info about each dataset for later use */
```

Comment:
In a few of the functions below, this is a common pattern. Whereas the parallel compression code previously worked on a single dataset at a time, the multi-dataset case means that the code needs a quick way to find the dataset that a particular chunk belongs to. If there's only one dataset in the I/O operation, this works almost exactly as before, with the exception of an extra small allocation to facilitate the multi-dataset I/O case. If multiple datasets are involved, a hash table is used to cache relevant info about each dataset, with the dataset info pointer or dataset's object header address (depending on the context) being used as the key.

Reply:
Maybe add this as a code comment?

Reply:
With some edits for context